# 📝 MARKER'S SUMMARY — 2024 A-Level Computing Paper 2 (marked against the supplied answer file)

| Question | Task marks | Score |
|---|---|---|
| Q1 Card game | 1.1 **5/5** · 1.2 **2/2** · 1.3 **4/8** · 1.4 **5/6** · 1.5 **2/3** | **18 / 24** |
| Q2 Number tree (OOP) | 2.1 **4/4** · 2.2 **3/3** · 2.3 **2/2** · 2.4 **3/3** · 2.5 **3/3** · 2.6 **3/4** · 2.7 **6/6** | **24 / 25** |
| Q3 HTML generator | 3.1 **2/2** · 3.2 **1/2** · 3.3 **5/6** · 3.4 **12/12** | **20 / 22** |
| Q4 Villa bookings (SQL) | 4.1 **6/7** · 4.2 **5/5** · 4.3 **5/5** · 4.4 **5/7** | **21 / 24** |
| **Total** | (marks as itemised in the answer file) | **83 / 95 (≈ 87 %)** |

**Where the marks went**
1. **Sort tie-break and return value (Task 1.3):** equal numbers aren't ordered red/green/blue (2 marks), the function returns `None` (1), hand output isn't on one line (1). The same one-card-per-line output also cost 1 mark in 1.4.
2. **`winner()` never called** (1.5): 1 mark.
3. **Logic bugs:** `start` offset in `create_number_tree()` (levels 4+ wrong, so the level-5 output is wrong): 1 mark; `insert(-2, …)` puts the elements outside the `<html><body>` tags and the closing tags are reversed in `body()`: 1 mark; available/unavailable swapped in `check_avail()`: 2 marks.
4. **Format slips:** URL not in quotes in `hyperlink()` (1), `customerName INTEGER` where the data is a customer ID (1).

**Python 3.7:** your saved outputs were produced by a much newer Python than 3.7 (the code uses `date.strptime`, which only exists in the newest releases, and a backslash inside an f-string). On 3.7, two cells would not run:
- **Task 4.3:** `date.strptime` raises `AttributeError`; nothing in the cell executes.
- **Task 4.4:** `'\n'.join(...)` inside an f-string is a `SyntaxError`; the function is never defined.

I marked the code as run on your machine; if the exam environment is 3.7, **up to 10 of the marks in Task 4.3 and 4.4 are at risk**. Both fixes are one-liners (see those tasks).

**Readability (overall):** names are mostly clear (`calculate_nodes`, `parse_insert`, `p1_colour`) and the generic helpers (`parse_insert`, the element → function dictionary) are a real strength. Biggest wins: replace one-card-per-line output with `print("Label:", hand)`; split `game()` into set-up / calculate / print; use `_` for unused loop variables and move imports to the top; never reuse one name for two things (`i`, `start_date`, `winner`); build strings in a variable before putting them in an f-string; and **run every cell and save**: the winner function, hyperlink quoting and the level-5 output would all have been caught by one more look at the output.

*Marks are my own application of the supplied marking points; I re-ran your code on stand-in data files (the `resources/*.txt` files were not supplied) and cross-checked scoring and the number tree against independent reference code.*


### 🐍 Python 3.7 compatibility scan (this notebook)

| Cell / task | Line | Problem on 3.7 | Fix |
|---|---|---|---|
| Task 4.3 | `date.strptime(f"{entry[-2]}-2025", "%d-%b-%Y")` | `AttributeError`: `date` has no `strptime` until the newest Python releases | `from datetime import datetime, timedelta` → `datetime.strptime(...)` |
| Task 4.4 | `print(f"""… {'\n'.join([...])} …""")` | `SyntaxError`: backslash inside an f-string `{ }` (allowed only from 3.12) | `avail_text = "\n".join(...)` on its own line, then `{avail_text}` |
| Task 4.4 | `date.strptime(...)` again | same as 4.3 | same as 4.3 |

Nothing else in this notebook uses a post-3.7 feature. (In the 2025 paper, `parse_insert` reused double quotes inside an f-string: also a `SyntaxError` before 3.12.)

**Common syntax/features that do not exist in Python 3.7**
- 3.8: walrus `:=`; positional-only parameters (`def f(a, /, b)`); `f"{x=}"`; unparenthesised `return a, *b`; `math.prod`, `math.isqrt`.
- 3.9: dict merge `a | b`; built-in generics `list[int]`; `str.removeprefix()` / `removesuffix()`.
- 3.10: `match` / `case`; `int | None` in annotations; `zip(..., strict=True)`; `itertools.pairwise`.
- 3.11: `except*` / `ExceptionGroup`; `tomllib`.
- 3.12: backslashes, comments and reused quotes inside f-string `{ }`; `type X = ...`; `def f[T](x)`.
- Available in 3.7: f-strings (`f"{x}"`), `dataclasses`, guaranteed dict insertion order, `from __future__ import annotations`.


# [ALVL/2024/ALVL/P2/Q1]
A game designer is designing a game for 2 players and needs you to program a prototype.

Players in the game have 10 randomly generated cards. This is called a 'hand'. The cards each have a colour (red, green or blue) and a number (1, 2, 3 or 4). There is no limit to the number of copies of each card, for example there can be multiple cards that are red and 1. Each player has a score that is 0 when the game starts.

In this prototype, the cards will be randomly generated for each player and sorted into ascending order by number.

On each turn:
- The next card from Player 1 's hand is selected.
- The next card from Player 2's hand is selected.
- These 2 cards are compared.

The winning card is determined based on the colour of each card:
- A red card wins over a green card.
- A green card wins over a blue card.
- A blue card wins over a red card.

The number on the winning card is multiplied by 2. This value is then added to the score of the player who had the winning card.

The number on the losing card is added to the score of the player who had the losing card.

If both colours are the same:
- Player 1 's card number is added to player 1 's score.
- Player 2's card number is added to player 2's score.

## Task 1.1
Write a program to set up the game for 2 players by:
- creating a list for each player
- randomly generating 10 cards (each with a colour and number) to each player's list.
<div style="text-align: right">[4]</div> 

Test your program by outputting the contents of each player's hand. <div style="text-align: right">[1]</div> 

In [1]:
# Task 1.1

def setup():
    import random
    players = []
    colours = ["red","green","blue"]

    for _ in range(2):
        hand = [(random.choice(colours),random.randint(1,4)) for i in range(10)]
        players.append(hand)

    return players

p1,p2 = setup()
print("Player 1's Hand:")
print(*p1,sep="\n")
print("\nPlayer 2's Hand:")
print(*p2,sep="\n")    

Player 1's Hand:
('blue', 4)
('green', 4)
('red', 2)
('red', 3)
('blue', 2)
('red', 2)
('blue', 4)
('red', 2)
('red', 3)
('red', 4)

Player 2's Hand:
('green', 3)
('red', 1)
('green', 1)
('red', 3)
('green', 1)
('green', 3)
('red', 1)
('blue', 3)
('red', 3)
('red', 3)


### 📝 MARKING — Task 1.1 setup(): **5 / 5**

| # | Marking point | Mark |
|---|---|---|
| 1 | Suitable list for Player 1's hand | ✅ |
| 2 | Suitable list for Player 2's hand | ✅ |
| 3 | Randomly generates valid colours (red, green, blue) | ✅ |
| 4 | Randomly generates valid numbers 1–4 | ✅ `randint(1,4)` |
| 5 | Outputs both hands for testing | ✅ |

**Comments**

- Compact and correct: a list comprehension of `(colour, number)` tuples, two hands from one loop.
- **Readability:** `for i in range(10)` never uses `i`; write `_`. `import random` belongs at the top of the file, not inside the function.
- The 4-line print block is copy-pasted three times in Q1. A `show_hand(name, hand)` helper that prints a hand on **one line** would remove the repetition and also fix a requirement in 1.3.

## Task 1.2
Write a function that takes the colour of two cards as parameters and returns the result of the comparison. The function should return '1' if the first colour wins, '2' if the second colour wins and '3' if the colours are the same. <div style="text-align: right">[2]</div> 

In [2]:
# Task 1.2

def winning_card(colour_1,colour_2):
    if colour_1 == colour_2:
        return 3
    colour_1_win_condition = [
        ("blue","red"),
        ("green","blue"),
        ("red","green")
    ]

    if (colour_1,colour_2) in colour_1_win_condition:
        return 1
    else:
        return 2

### 📝 MARKING — Task 1.2 compare colours: **2 / 2**

| # | Marking point | Mark |
|---|---|---|
| 1 | Identifies when the first colour beats the second | ✅ |
| 2 | Identifies when the second colour wins, or both are the same | ✅ |

**Comments**

- Clean: a list of winning `(first, second)` pairs is easy to check against the paper.
- The paper asks for `'1'`, `'2'`, `'3'` (strings); you return the integers `1`, `2`, `3`. I accepted this, but a strict marker could object. Returning `"1"` costs nothing.
- The name `winning_card` is misleading (it compares *colours* and can return a draw); `compare_colours` says what it does. There is no test of this function (the reference tests all 9 colour pairs): not a mark point, but a 3-line loop would prove it.

## Task 1.3
The cards in each hand now need to be sorted into numerical order.

Write a function to take a hand as a parameter, sort the hand in-place into ascending numerical order and return the sorted hand. If two cards have the same number, the colours are compared to position the cards in this order: red, green, blue.

Call the function to sort each of the 2 players' hands. Output the contents of each hand before and after sorting. The contents of a hand must be output on one line.

Do **not** use a built-in sorting function. <div style="text-align: right">[8]</div> 

In [3]:
def insertion_sort(hand):
    for i in range(len(hand)):
        j = i
        while j > 0 and hand[j-1][1] > hand[j][1]:
            hand[j-1],hand[j] = hand[j],hand[j-1]
            j -= 1

insertion_sort(p1)
insertion_sort(p2)

print("Player 1's Hand:")
print(*p1,sep="\n")
print("\nPlayer 2's Hand:")
print(*p2,sep="\n")    

Player 1's Hand:
('red', 2)
('blue', 2)
('red', 2)
('red', 2)
('red', 3)
('red', 3)
('blue', 4)
('green', 4)
('blue', 4)
('red', 4)

Player 2's Hand:
('red', 1)
('green', 1)
('green', 1)
('red', 1)
('green', 3)
('red', 3)
('green', 3)
('blue', 3)
('red', 3)
('red', 3)


### 📝 MARKING — Task 1.3 sort the hands: **4 / 8**

| # | Marking point | Mark |
|---|---|---|
| 1 | Defines a function that accepts a hand | ✅ |
| 2 | Does not use a built-in sort | ✅ insertion sort |
| 3 | Sorts by number in ascending order **[2]** | ✅ 2/2 |
| 4 | Sorts equal numbers by colour order red, green, blue **[2]** | ❌ 0/2 — colour is never compared |
| 5 | Returns the sorted hand | ❌ the function returns `None` |
| 6 | Outputs both hands before and after sorting for testing | ❌ printed one card per line, not "on one line" as the question requires; the 'before' print is in another cell |

**Comments**

- **Tie-break missing.** The `while` test only compares `hand[j-1][1] > hand[j][1]` (the numbers). I tested `[blue 2, red 2, green 2, red 1]` and got `[red 1, blue 2, red 2, green 2]`: the colours are in no particular order.
- **Fix (also tidier):** give each colour a rank and compare `(number, rank)` pairs: `ORDER = {'red': 0, 'green': 1, 'blue': 2}` then `key = lambda c: (c[1], ORDER[c[0]])` and swap while `key(hand[j-1]) > key(hand[j])`. (The key function is only a comparison helper; the sort loop is still yours, so no built-in sort is used.) End the function with `return hand`.
- The outer loop can start at `1`: `range(len(hand))` makes one wasted pass.
- Output: `print("Before:", hand)` and `print("After:", hand)` print the whole hand on one line. `print(*p1, sep="\n")` puts each card on its own line.

## Task 1.4
The cards are compared in pairs, starting with the first card from player 1 compared to the first card from player 2. 

The appropriate numbers are added to each player's score until all of the 10 cards have been used.

Write a function to calculate and return the final score for each of the 2 players. <div style="text-align: right">[5]</div> 

Test your program by:
- outputting the sorted hand and the final score for player 1 on one line
- outputting the sorted hand and the final score for player 2 on one line. <div style="text-align: right">[1]</div> 

In [4]:
# Task 1.4

def game():
    p1_score = p2_score = 0
    player_1,player_2 = setup()
    insertion_sort(player_1)
    insertion_sort(player_2)
    print("Player 1's Hand(Sorted):")
    print(*player_1,sep="\n")
    print()
    print("Player 2's Hand(Sorted):")
    print(*player_2,sep="\n")
    print()

    for i in range(10):
        (p1_colour,p1_number),(p2_colour,p2_number) = player_1[i],player_2[i]
        winner = winning_card(p1_colour,p2_colour)
        if winner == 3:
            p1_score += p1_number
            p2_score += p2_number

        elif winner == 2:
            p1_score += p1_number
            p2_score += p2_number*2

        else:
            p1_score += p1_number * 2
            p2_score += p2_number

    print(f"Player 1's Final Score: {p1_score}")
    print(f"Player 2's Final Score: {p2_score}")

    return p1_score,p2_score

game()
            

Player 1's Hand(Sorted):
('red', 1)
('green', 1)
('green', 2)
('blue', 2)
('green', 3)
('green', 4)
('green', 4)
('green', 4)
('blue', 4)
('red', 4)

Player 2's Hand(Sorted):
('green', 1)
('red', 2)
('green', 2)
('red', 3)
('green', 3)
('blue', 3)
('green', 3)
('red', 3)
('blue', 3)
('red', 3)

Player 1's Final Score: 36
Player 2's Final Score: 31


(36, 31)

### 📝 MARKING — Task 1.4 scores: **5 / 6**

| # | Marking point | Mark |
|---|---|---|
| 1 | Defines a function to calculate the scores | ✅ (`game()`) |
| 2 | Compares corresponding cards from both hands | ✅ |
| 3 | Doubles the winning card's number for the winner | ✅ |
| 4 | Adds the losing card's number to the loser | ✅ |
| 5 | Handles both cards the same colour | ✅ |
| 6 | Outputs each sorted hand and its final score on the same line | ❌ hand is printed one card per line, score printed separately |

**Comments**

- Logic verified: I compared `game()` with an independent scorer over 300 random games, with **0 mismatches**.
- **Design:** `game()` creates new hands, sorts, prints and scores in one 30-line function, but the question asks for a function that *calculates and returns the final scores*. Splitting out `calculate_scores(hand1, hand2)` also means 1.5 can call it. It also silently throws away the hands you sorted in 1.3.
- **Readability:** the three branches repeat the same additions. One line each does the job: `p1_score += n1 * (2 if result == 1 else 1)` and `p2_score += n2 * (2 if result == 2 else 1)`.

## Task 1.5
Write a function to calculate and output which player has won (has the higher final score), and which player has come second.

There must be a suitable output if there is a draw.

All outputs must have an appropriate message. <div style="text-align: right">[2]</div> 

Test your program. <div style="text-align: right">[1]</div> 

In [5]:
# Task 1.5

def winner(p1_score,p2_score):
    if p1_score > p2_score:
        print(f"Player 1 wins with {p1_score} points.\nPlayer 2 comes in second with {p2_score}")
    elif p2_score > p1_score:
        print(f"Player 2 wins with {p2_score} points.\nPlayer 1 comes in second with {p1_score}")
    else:
        print(f"Player 1 and Player 2 tie, both scoring {p1_score} points")


### 📝 MARKING — Task 1.5 announce the winner: **2 / 3**

| # | Marking point | Mark |
|---|---|---|
| 1 | Identifies the higher score as the winner | ✅ |
| 2 | Handles and outputs a draw | ✅ |
| 3 | Tests the winner function and shows the result | ❌ `winner()` is defined but never called |

**Comments**

- Messages are clear and give both players' scores. Easy mark lost: add `winner(*game())` (or pass the scores returned by `calculate_scores`) and run the cell.
- `winner` is also the name of a local variable inside `game()` (the colour-comparison result). Different scopes, but it reads confusingly; call the function `announce_result`.

# [ALVL/2024/ALVL/P2/Q2]
A program is needed to calculate the numbers on each level of a number tree.

The bottom row of the number tree has the numbers 1 to x.

The row above has the result of the addition of the two numbers directly below it. This repeats in pairs until there is only one number left at the top of the number tree. This creates the number tree.

Example 1: This number tree has 2 levels. The first level has the numbers 1 to 2.

The next level adds together these numbers to give 3. There is only 1 number on this level, so the number tree is complete.

<center>

<img src="img/2024_alvl_p2_q2_img_1.png" width="600" align="center"/>

</center>

Example 2: This number tree has 3 levels. The first level has the numbers 1 to 4.

The second level adds together these numbers in pairs to give 1 + 2 = 3 and 3 + 4 = 7.

The third level adds together the pair on the second level, 3 + 7 = 10.

There is now only one number at this level, so the number tree is complete.

<center>

<img src="img/2024_alvl_p2_q2_img_2.png" width="600" align="center"/>

</center>

The number of levels is input into the program.

The program will need to generate and output the number tree for the given number of levels.

The program uses object-oriented programming. Each number in the number tree is a node. The nodes are stored in a list. Each node has the integer number (for example, 1 ), the index of the left node below it and the index of the right node below it.

The program needs to use local variables throughout.

The class `Node` contains three attributes:
- `data_value` the node's integer data, initialised to -1
- `left_node` the index of the number below it to the left, initialised to -1
- `right_ node` the index of the number below it to the right, initialised to -1

-1 represents no data present in `data_value`

-1 represents a null value in a pointer.

The class `Node` has the following methods:
- a constructor that initialises the three attributes
- `set_data()` that assigns a value to `data_value`
- `set_left()` that assigns a value to `left_node`
- `set_right()` that assigns a value to `right_node`
- `get_data()` that returns `data_value`
- `get_left()` that returns `left_node`
- `get_right()` that returns `right_node`

## Task 2.1 
Write program code to declare the class `Node`, its constructor and its methods. <div style="text-align: right">[4]</div> 

In [6]:
# Task 2.1
class Node:
    def __init__(self):
        self.data_value = -1
        self.left_node = -1
        self.right_node = -1

    def set_data(self,val):
        self.data_value = val

    def set_left(self,val):
        self.left_node = val

    def set_right(self,val):
        self.right_node = val

    def get_data(self):
        return self.data_value

    def get_left(self):
        return self.left_node

    def get_right(self):
        return self.right_node

        

### 📝 MARKING — Task 2.1 class Node: **4 / 4**

| # | Marking point | Mark |
|---|---|---|
| 1 | Defines a suitable Node class | ✅ |
| 2 | Constructor initialises the three attributes to -1 | ✅ |
| 3 | Correct setters for data, left, right | ✅ |
| 4 | Correct getters for data, left, right | ✅ |

**Comments**

- Fully correct and matches the paper's method list exactly (no extras this time, which is good).
- Trailing blank lines at the end of the cell; add a one-line class comment saying `-1` means "no data / no pointer".

## Task 2.2
This table shows the total number of nodes for number trees with levels 1 to 5.

<center>

| Number of levels | Total number of nodes |
|------------------|-----------------------|
| 1                | 1                     |
| 2                | 3                     |
| 3                | 7                     |
| 4                | 15                    |
| 5                | 31                    |

</center>

The function `calculate_nodes()` takes the number of levels as a parameter. It calculates, outputs and returns the number of nodes for that number tree.

The output must be in an appropriate message, for example:

`A 3 level number tree has 7 nodes`

You do not need to validate the parameter. The function must calculate the number of nodes for any number of levels greater than 0.

Write the function `calculate_nodes()`. <div style="text-align: right">[3]</div> 

In [7]:
# Task 2.2

def calculate_nodes(number_of_levels):
    number_of_nodes = 2**(number_of_levels)-1 if number_of_levels > 1 else 1
    print(f"A {number_of_levels} level number tree has {number_of_nodes} nodes.")
    return number_of_nodes

### 📝 MARKING — Task 2.2 calculate_nodes(): **3 / 3**

| # | Marking point | Mark |
|---|---|---|
| 1 | Accepts the number of levels | ✅ |
| 2 | Calculates the total number of nodes as 2**levels − 1 | ✅ |
| 3 | Outputs an appropriate message and returns the value | ✅ |

**Comments**

- The conditional `if number_of_levels > 1 else 1` is unnecessary: `2**1 − 1` is already `1`. Delete it so the formula is the whole function.
- The message ends with a full stop and says `1 nodes` for one level; the paper's example has no full stop.

## Task 2.3
Test your program four times, with the following levels:

```
1
3
5
10
```

Show the output for each test. <div style="text-align: right">[2]</div> 

In [8]:
print(calculate_nodes(1))
print(calculate_nodes(3))
print(calculate_nodes(5))
print(calculate_nodes(10))

A 1 level number tree has 1 nodes.
1
A 3 level number tree has 7 nodes.
7
A 5 level number tree has 31 nodes.
31
A 10 level number tree has 1023 nodes.
1023


### 📝 MARKING — Task 2.3 test calculate_nodes(): **2 / 2**

| # | Marking point | Mark |
|---|---|---|
| 1 | Tests levels 1, 3, 5 and 10 | ✅ |
| 2 | Shows the output clearly for each test | ✅ output saved |

**Comments**

- Output is saved and correct (1, 7, 31, 1023). Each level prints the number twice (once from the function, once from `print(...)`); harmless, but just call `calculate_nodes(1)` since it already prints.

## Task 2.4
The function `create_empty_tree()` takes 2 parameters; the list of nodes and the number of levels.

The function creates a list of `Node` objects in their initial state with all attributes set to -1. The function returns the list.

The list is zero indexed.

Write the function `create_empty_tree()`. <div style="text-align: right">[3]</div>

In [9]:
# Task 2.4

def create_empty_tree(list_of_nodes,number_of_levels):
    number_of_nodes = calculate_nodes(number_of_levels)
    list_of_nodes = [Node() for i in range(number_of_nodes)]

    return list_of_nodes

list_of_nodes = create_empty_tree([],3)

A 3 level number tree has 7 nodes.


### 📝 MARKING — Task 2.4 create_empty_tree(): **3 / 3**

| # | Marking point | Mark |
|---|---|---|
| 1 | Defines the function with suitable parameters | ✅ |
| 2 | Calculates the correct total number of nodes | ✅ via `calculate_nodes()` |
| 3 | Returns a list of the correct number of empty Nodes | ✅ |

**Comments**

- Correct. The `list_of_nodes` parameter is immediately overwritten, as in the reference: fine, but a short comment saying why it is unused avoids confusion.
- Local variable names `number_of_nodes` / `list_of_nodes` are descriptive; consider `count` and `nodes` to keep lines short.

## Task 2.5
The function `assign_leaf_nodes()` sets the data for the lowest level of nodes. The function takes 2 parameters; the list of nodes and the number of levels.

Each leaf node in the first layer is assigned a number starting with 1 in the left-most node and then incrementing by 1 for each subsequent node.

For example, the first level of nodes in this number tree contain the data 1 2 3 4

<center>

<img src="img/2024_alvl_p2_q2_img_2.png" width="600" align="center"/>

</center>

This table shows the number of leaf nodes for up to 5 levels in a number tree. The function must calculate the number of leaf nodes for any number of levels greater than 0.

<center>

| Number of levels | Number of leaf nodes  |
|------------------|-----------------------|
| 1                | 1                     |
| 2                | 2                     |
| 3                | 4                     |
| 4                | 8                     |
| 5                | 16                    |

</center>

For example, if the list of nodes is named nodes, then `assign_leaf_nodes(nodes, 3)` would return a list of 7 `Node` objects with the following attributes:

<center>

<img src="img/2024_alvl_p2_q2_img_3.png" width="600" align="center"/>

</center>

Write the function `assign_leaf_nodes()`. <div style="text-align: right">[3]</div>

In [10]:
# Task 2.5

def assign_leaf_nodes(list_of_nodes,number_of_levels):
    number_of_leaf_nodes = 2**(number_of_levels - 1)

    for i in range(number_of_leaf_nodes):
        list_of_nodes[i].set_data(i+1)

    return list_of_nodes


list_of_nodes = assign_leaf_nodes(list_of_nodes,3)
print([i.get_data() for i in list_of_nodes])

[1, 2, 3, 4, -1, -1, -1]


### 📝 MARKING — Task 2.5 assign_leaf_nodes(): **3 / 3**

| # | Marking point | Mark |
|---|---|---|
| 1 | Calculates the number of leaf nodes as 2**(levels−1) | ✅ |
| 2 | Assigns leaf data starting from 1 | ✅ |
| 3 | Returns the updated list | ✅ |

**Comments**

- Correct, and the saved output `[1, 2, 3, 4, -1, -1, -1]` matches the paper's diagram.

## Task 2.6
The data for each parent node in the number tree is calculated by adding the data from its two lower nodes.

The function `create_number_tree()`:
- takes 2 parameters: the list of nodes (with the first level of nodes already assigned their data) and the number of levels
- calculates the data for each node in each level in the number tree
- stores the data in each node
- assigns the left pointer of each node to its left lower node
- assigns the right pointer of each node to its right lower node.

For example, if the list of nodes is named `nodes`, then `create_number_tree(nodes, 3)` would return a list of 7 `Node` objects with the following attributes:

<center>

<img src="img/2024_alvl_p2_q2_img_4.png" width="600" align="center"/>

</center>

Write the function `create_number_tree()`. <div style="text-align: right">[4]</div>

In [ ]:
# Task 2.6

def create_number_tree(list_of_nodes,number_of_levels):
    lvl_number = number_of_levels
    start = 0
    number_of_nodes_in_lvl = 2**(lvl_number - 1)
    j = number_of_nodes_in_lvl
    
    while j < len(list_of_nodes):
        for i in range(0,number_of_nodes_in_lvl,2):
            
            parent_data = sum([i.get_data() for i in list_of_nodes[start+i:start+i+2]])
            list_of_nodes[j].set_data(parent_data)
            list_of_nodes[j].set_left(start+i)
            list_of_nodes[j].set_right(start+i+1)
            j+=1

        start += number_of_nodes_in_lvl
        lvl_number -= 1
        number_of_nodes_in_lvl = 2**(lvl_number - 1)

    return list_of_nodes


list_of_nodes = create_number_tree(list_of_nodes,3)

for i in list_of_nodes:
    print(i.get_data(),i.get_left(),i.get_right())

1 -1 -1
2 -1 -1
3 -1 -1
4 -1 -1
3 0 1
7 2 3
10 4 5


### 📝 MARKING — Task 2.6 create_number_tree(): **3 / 4**

| # | Marking point | Mark |
|---|---|---|
| 1 | Correctly processes each non-leaf level | ❌ see bug: `start` is not cumulative, so levels 4+ read the wrong nodes |
| 2 | Parent value = sum of its two children | ✅ formula is right |
| 3 | Assigns the left pointer of each parent | ✅ `start+i` |
| 4 | Assigns the right pointer of each parent | ✅ `start+i+1` |

**Comments**

- **Bug:** at the end of each level you write `start = number_of_nodes_in_lvl`. It should be `start += number_of_nodes_in_lvl` (running total of nodes already used). It happens to work for 1–3 levels, which is the paper's example, but fails from 4 levels. I compared against an independent builder: levels 1–3 correct; level 4 first wrong at index 14 (`(4, 11, 5)` instead of `(12, 36, 13)`); level 5 wrong from index 28. **Your saved level-5 output in 2.7 is therefore wrong.**
- **Readability:** the comprehension variable `i` shadows the loop variable `i` in the same line: `sum([i.get_data() for i in list_of_nodes[start+i:start+i+2]])`. Python 3 happens to evaluate the slice in the outer scope so it runs, but it is a trap. Use `left = start + i` and `list_of_nodes[left].get_data() + list_of_nodes[left+1].get_data()`.
- `lvl_number` and `number_of_nodes_in_lvl` track the same thing; halving one variable (`width //= 2`) is simpler.

## Task 2.7
Amend the program to call the functions `calculate_nodes()`, `create_empty_tree()`, `assign_leaf_nodes()` and `create_number_tree()`.

Output the content of each node in the array in the format:

<center>

left_pointer    data    right_pointer

</center>

<div style="text-align: right">[4]</div>

Test your program with levels 1, 3 and 5 and show the output for each test. <div style="text-align: right">[2]</div>

In [12]:
# Task 2.7


def main(number_of_levels):
    list_of_nodes = create_empty_tree([],number_of_levels)
    list_of_nodes = assign_leaf_nodes(list_of_nodes,number_of_levels)
    list_of_nodes = create_number_tree(list_of_nodes,number_of_levels)

    for i in list_of_nodes:
        print(f"{i.get_left():<2} {i.get_data():<2} {i.get_right():<2}") 

print("Number tree with 1 level")
main(1)
print("\nNumber tree with 3 levels")
main(3)
print("\nNumber tree with 5 levels")
main(5)

Number tree with 1 level
A 1 level number tree has 1 nodes.
-1 1  -1

Number tree with 3 levels
A 3 level number tree has 7 nodes.
-1 1  -1
-1 2  -1
-1 3  -1
-1 4  -1
0  3  1 
2  7  3 
4  10 5 

Number tree with 5 levels
A 5 level number tree has 31 nodes.
-1 1  -1
-1 2  -1
-1 3  -1
-1 4  -1
-1 5  -1
-1 6  -1
-1 7  -1
-1 8  -1
-1 9  -1
-1 10 -1
-1 11 -1
-1 12 -1
-1 13 -1
-1 14 -1
-1 15 -1
-1 16 -1
0  3  1 
2  7  3 
4  11 5 
6  15 7 
8  19 9 
10 23 11
12 27 13
14 31 15
16 10 17
18 26 19
20 42 21
22 58 23
8  19 9 
10 23 11
4  11 5 


### 📝 MARKING — Task 2.7 main() and tests: **6 / 6**

| # | Marking point | Mark |
|---|---|---|
| 1 | Creates the required number of nodes | ✅ |
| 2 | Calls create_empty_tree() | ✅ |
| 3 | Calls assign_leaf_nodes() | ✅ |
| 4 | Calls create_number_tree() | ✅ |
| 5 | Outputs each node as left_pointer data right_pointer | ✅ |
| 6 | Tests levels 1, 3 and 5 | ✅ |

**Comments**

- All six points are met as worded. However, the level-5 tree you display is wrong (see 2.6): check results against the paper's pattern (top node of a 5-level tree is 136, yours shows 11).
- `main()` is a good name for a driver. Printing a title line before each test (`Number tree with 3 levels`) makes the output easy to read: nice.

# [ALVL/2024/ALVL/P2/Q3]

Write a program to format a description for a webpage into an HTML document.

The description for the program will be provided in a text file. Each element will be on a new line and will follow the structure:

`HTML element: content`

For example, in the text file `firstPage.txt` the first line is:

heading: My first page

The HTML element is 'heading' and the text for the heading is 'My first page'.

The HTML elements that can be included are in this table with an example and a description for the content.

<center>

| HTML Element | Example                                   | Description                                                            |
|--------------|-------------------------------------------|------------------------------------------------------------------------|
| heading      | `heading: This is a title`                  | Formats the string `"This is a title"` with the h1 tag                   |
| subheading   | `subheading: This is a subtitle`            | Formats the string `"This is a subtitle"` with the h2 tag                |
| hyperlink    | `hyperlink: https://www.seab.gov.sg, Click` | Formats the string `"Click"` as a hyperlink to `"https://www.seab.gov.sg"` |
| paragraph    | `paragraph: This is a paragraph`            | Formats the string `"This is a paragraph"` as a paragraph                |

</center>

All documents start and end with the HTML and body tags.

The data being used will not include any special characters. You will not be required to validate the content in each statement, for example that a title is a valid title.

## Task 3.1
Write **one** function for each of the HTML elements heading, subheading and paragraph to:
- take the text to format as a parameter
- format the text including HTML tags
- return the formatted text.

For example, if the text parameter for the heading function is "My first page", the formatted text will be:

`"<hl>My first page</hl >"` <div style="text-align: right">[1]</div>

Test your program by calling the three functions with `"My element"` and outputting the return value from each. <div style="text-align: right">[1]</div>

In [13]:
# Task 3.1

def heading(text):
    string = f"<h1>{text}</h1>"
    return string

def subheading(text):
    string = f"<h2>{text}</h2>"
    return string

def paragraph(text):
    string = f"<p>{text}</p>"
    return string

print(heading("My element"))
print(subheading("My element"))
print(paragraph("My element"))

<h1>My element</h1>
<h2>My element</h2>
<p>My element</p>


### 📝 MARKING — Task 3.1 heading / subheading / paragraph: **2 / 2**

| # | Marking point | Mark |
|---|---|---|
| 1 | Defines functions for heading, subheading, paragraph | ✅ |
| 2 | Tests all three with "My element" | ✅ output saved |

**Comments**

- Correct. The temporary `string` variable is unnecessary: `return f"<h1>{text}</h1>"` is one line.

## Task 3.2
Write a function for the HTML element hyperlink to:
- take the text to format as a parameter
- split the parameter into the URL and string
- format the text including HTML tags
- return the formatted text.

For example, if the text parameter is `"https: / /www. cambridge. org, Click here"`, the formatted text will be:

`"<a href="https://www.cambridge.org" >Click here</a>"` <div style="text-align: right">[1]</div>

Test your program by calling the function with `"https://www.cambridgeinternational.org, Click here"` and outputting the return value. <div style="text-align: right">[1]</div>

In [ ]:
# Task 3.2

def hyperlink(text):
    # Assumes the text and link both do not contain any commas
    link,string = [i.strip() for i in text.split(",")]
    formatted_string = f"<a href=\"{link}\" >{string}</a>"
    return formatted_string

print(hyperlink("https://www.cambridgeinternational.org, Click here"))
    

<a href=https://www.cambridgeinternational.org >Click here</a>


### 📝 MARKING — Task 3.2 hyperlink(): **1 / 2**

| # | Marking point | Mark |
|---|---|---|
| 1 | Separates the URL and the display text | ✅ `split(",")` + `strip()` |
| 2 | Returns the hyperlink in correct anchor-tag format and tests it | ❌ `href` value has no quotes |

**Comments**

- Your output is `<a href=https://www.cambridgeinternational.org >Click here</a>`. The paper's example is `<a href="https://www.cambridge.org" >Click here</a>`: the URL must be in quotes. Browsers tolerate unquoted attributes when the URL has no spaces, `=` or `&`, but a URL like `...?id=5` would break it, so quote it.
- Good: you strip the space after the comma, and you commented the assumption that neither part contains a comma.

## Task 3.3
The function `body()` takes a list as a parameter containing the text to convert into an HTML formatted document. Each element in the list has one HTML element for conversion.

Write the function to:
- call the appropriate function for each HTML element in the list
- create a single string containing all the formatted returned HTML elements
- insert the HTML and body opening and closing tags at the start and end of the string
- return either the complete formatted string, or False if there are any unrecognised elements. <div style="text-align: right">[6]</div>

In [ ]:
# Task 3.3

def body(txt_to_convert_lst):
    full_string_lst = ["<html><body>","</body></html>"]
    element_map_dict = {
        "heading":heading,
        "subheading":subheading,
        "hyperlink":hyperlink,
        "paragraph":paragraph
    }
    for element,text in txt_to_convert_lst:
        if element not in element_map_dict:
            return False
        else:
            formatted_element = element_map_dict[element](text)
            full_string_lst.insert(-2,formatted_element)

    return "".join(full_string_lst)
            

### 📝 MARKING — Task 3.3 body(): **5 / 6**

| # | Marking point | Mark |
|---|---|---|
| 1 | Iterates through each element in the list | ✅ |
| 2 | Processes heading elements | ✅ |
| 3 | Processes subheading elements | ✅ |
| 4 | Processes paragraph elements | ✅ |
| 5 | Processes hyperlink elements | ✅ |
| 6 | Adds the opening/closing HTML and body tags and returns False for unrecognised elements | ❌ tags are in the wrong place and wrong order (False is correct) |

**Comments**

- **Bug:** `full_string_lst.insert(-2, element)` on a list that starts as `[open_tags, close_tags]` inserts *before the opening tags*. Output for my test file: `<h1>…</h1><h2>…</h2>…<html><body></html></body>`. The document should be `<html><body>…</body></html>`.
- **Also:** the closing tags are `</html></body>` (reversed). Body closes first.
- **Fix (simpler):** collect the converted elements in a list, then `return "<html><body>" + "".join(parts) + "</body></html>"`. No `insert`, no index arithmetic.
- Nice: the dictionary mapping element name → function removes an if/elif chain and handles the unknown-element case in one test.

## Task 3.4
The text file `firstPage.txt` contains the description of a web page that needs converting into an HTML document.

Write the function `read_data()`, to read in the data from a text file. The function needs to:
- take the name of a file (including extension) as input
- read each line from the file into a separate list element
- call the appropriate functions to convert the list to a formatted HTML string
- return the formatted HTML string.

Write the function `write_data()`, to store the formatted HTML string as an HTML file. The function needs to:
- take the name of a file (excluding extension) as input
- take the formatted HTML string to write as input
- insert the string `<!doc type html>` at the start of the file
- write the formatted string to the file with the name input and the extension `".html"`.

Output an appropriate message if the data cannot be converted into HTML.

Call the function `read_data()` then call the function `write_data()`. <div style="text-align: right">[9]</div>

Run your program two times with the following inputs:

Test 1:

`firstPage.txt`

`test`

Test 2:

`website.txt`

`theWebsite` <div style="text-align: right">[3]</div>

In [16]:
# Task 3.4

def read_data(filename):
    with open(f"resources/{filename}") as f:
        data = [[j.strip() for j in i.split(":",1)] for i in f.readlines()]

    formatted_string = body(data)
    return formatted_string

def write_data(filename,formatted_string):
    full_string = "<!DOCTYPE html>" + formatted_string
    with open(f"{filename}.html","w") as f:
        f.write(full_string)


formatted_string = read_data("firstPage.txt")
if formatted_string:
    write_data("test",formatted_string)
else:
    print("Unknown element detected when parsing text to html")

formatted_string = read_data("website.txt")
if formatted_string:
    write_data("theWebsite",formatted_string)
else:
    print("Unknown element detected when parsing text to html")

### 📝 MARKING — Task 3.4 read_data(), write_data(), tests: **12 / 12**

| # | Marking point | Mark |
|---|---|---|
| 1 | read_data() accepts a filename including extension | ✅ |
| 2 | Opens and reads each line of the input file | ✅ |
| 3 | Stores each line as a separate list element | ✅ |
| 4 | Splits each line into element type and content | ✅ `split(":", 1)` keeps `https://` intact |
| 5 | Calls body() | ✅ |
| 6 | Returns the formatted HTML string | ✅ |
| 7 | write_data() accepts the output filename without extension | ✅ |
| 8 | write_data() accepts the HTML string | ✅ |
| 9 | Inserts the document-type declaration at the start | ✅ `<!DOCTYPE html>` |
| 10 | Writes to `<name>.html` | ✅ |
| 11 | Outputs an error message if conversion fails | ✅ in the calling code |
| 12 | Tests firstPage.txt → test and website.txt → theWebsite | ✅ |

**Comments**

- All points met. (The paper's `<!doc type html>` is a misprint; `<!DOCTYPE html>` is correct and the reference uses the same.) The structure of the file-writing is sound, but see the 3.3 bug: the HTML written to the file is mis-ordered.
- **Robustness:** a blank line in the input (e.g. a trailing empty line) makes `body()` crash with `ValueError: not enough values to unpack`. Skip blank lines: `for line in f: if line.strip(): …`.
- The two test blocks are identical apart from the filenames; a helper `convert(infile, outfile)` or a loop over the two pairs would halve the code. Put the error message in the helper, so it is only written once.

# [ALVL/2024/ALVL/P2/Q4]
A holiday company has 10 villas that customers can hire. A program is being created to store the customer bookings and allow the company staff to check the availability of the villas on requested dates.

The text files `customerBookings.txt` and `villas.txt` store the current data.

The data in `villas.txt` is stored in the following format:

villa ID, villa name, country, cost

The data in customerBookings. txt is stored in the following format:

booking ID, customer ID, villa ID, start date, number of days

You do not need to consider how data about the customers will be stored.

You can assume the year is 2025.

## Task 4.1
Write a Python program that uses SQL to create a database with two tables: one table to store data about the villas and one table to store data about customer bookings. 

Define the primary and foreign keys for each table. <div style="text-align: right">[7]</div>

In [ ]:
def tables():
    import sqlite3
    tables_queries = [
        """
        CREATE TABLE IF NOT EXISTS Villa(
        villaID INTEGER PRIMARY KEY,
        villaName TEXT,
        villaCountry TEXT,
        villaCost INTEGER
        )
        """,
        """
        CREATE TABLE IF NOT EXISTS CustomerBooking(
        bookingID INTEGER PRIMARY KEY,
        customerName TEXT,
        villaID REFERENCES Villa(villaID),
        startDate TEXT,
        numberOfDays INTEGER
        )
        """
    ]

    delete_queries = ["DELETE FROM Villa","DELETE FROM CustomerBooking"]

    with sqlite3.connect("holiday_company.db") as con:
        c = con.cursor()
        c.executescript(";".join(tables_queries))
        c.executescript(";".join(delete_queries))

        con.commit()

tables()

### 📝 MARKING — Task 4.1 create the tables: **6 / 7**

| # | Marking point | Mark |
|---|---|---|
| 1 | Creates / connects to the SQLite database | ✅ |
| 2 | Creates the Villa table | ✅ |
| 3 | Creates the CustomerBooking table | ✅ |
| 4 | Appropriate fields for Villa | ✅ |
| 5 | Appropriate fields for CustomerBooking | ❌ `customerName INTEGER`: the file holds a customer **ID** |
| 6 | Correct primary keys | ✅ |
| 7 | Correct foreign key from CustomerBooking to Villa | ✅ `villaID REFERENCES Villa(villaID)` |

**Comments**

- `customerBookings.txt` stores *booking ID, customer ID, villa ID, start date, number of days*. Your column is named `customerName`, and typed `INTEGER`, which is both a naming and a typing error (IDs like `C01` would not be integers). Name it `customerID`.
- `villaID REFERENCES Villa(villaID)` has no data type; it works in SQLite but state it (`villaID INTEGER REFERENCES …`).
- As flagged in the 2025 paper, `tables()` also **deletes every row** each time it runs; keep creation and clearing separate.

## Task 4.2
Write a Python program that uses SQL to insert the data from each of the two text files into the appropriate field in each database table. <div style="text-align: right">[4]</div>

Run your program to test the data has been entered correctly. <div style="text-align: right">[1]</div>

In [18]:
tables()
def parse_insert(filename,tablename):
    import csv
    import sqlite3
    
    with open(f"resources/{filename}") as f:
        data = [tuple(i) for i in csv.reader(f)]

    with sqlite3.connect("holiday_company.db") as con:
        c = con.cursor()
        columns = len(data[0])
        placeholders = ",".join(["?"]*columns)
        c.executemany(f"INSERT INTO {tablename} VALUES ({placeholders})",data)

    return data

villa_data = parse_insert("villas.txt","Villa")
customerbooking_data=parse_insert("customerBookings.txt","CustomerBooking")

# Checking if data has been entered correctly
import sqlite3
with sqlite3.connect("holiday_company.db") as con:
    c = con.cursor()
    c.execute("SELECT * FROM Villa")
    check_data = c.fetchall()
    print("Villa Data:")
    print(*check_data,sep="\n")
    print()

    c.execute("SELECT * FROM CustomerBooking")
    check_data = c.fetchall()
    print("CustomerBookings Data:")
    print(*check_data,sep="\n")
    print()
    

Villa Data:
(1, 'Rose', 'France', 128)
(2, 'Sea view', 'Australia', 325)
(3, 'Dolphin', 'New Zealand', 490)
(4, 'Flower haven', 'Mexico', 580)
(5, 'Mountain breeze', 'India', 268)
(6, 'Sunset', 'UK', 136)
(7, 'Moonlight', 'USA', 358)
(8, 'White brick', 'Italy', 410)
(9, 'Blue house', 'Germany', 400)
(10, 'Walled garden', 'Croatia', 258)

CustomerBookings Data:
(1, 857, '3', '05-Jan', 7)
(2, 1149, '10', '20-Jan', 11)
(3, 388, '9', '05-Feb', 2)
(4, 230, '4', '03-Mar', 14)
(5, 1254, '6', '19-Nov', 10)
(6, 1500, '4', '12-May', 4)
(7, 1687, '7', '03-Mar', 6)
(8, 1408, '2', '18-Aug', 2)
(9, 1138, '3', '02-Mar', 4)
(10, 235, '5', '04-Apr', 7)
(11, 4, '10', '07-Jul', 12)
(12, 1981, '8', '15-Jan', 5)
(13, 500, '9', '06-Nov', 4)
(14, 332, '9', '03-Apr', 7)
(15, 1512, '5', '12-Oct', 14)
(16, 660, '4', '08-Jul', 6)
(17, 738, '4', '09-Aug', 7)
(18, 1211, '1', '12-Nov', 3)
(19, 1089, '1', '15-Sep', 4)
(20, 1294, '2', '03-Feb', 14)
(21, 1087, '3', '10-Apr', 14)
(22, 722, '3', '01-Feb', 2)
(23, 566, '

### 📝 MARKING — Task 4.2 insert + check: **5 / 5**

| # | Marking point | Mark |
|---|---|---|
| 1 | Reads data correctly from villas.txt | ✅ |
| 2 | Inserts villa records into the correct table | ✅ |
| 3 | Reads data correctly from customerBookings.txt | ✅ |
| 4 | Inserts booking records into the correct table | ✅ |
| 5 | Tests/displays the records to show they were inserted | ✅ SELECT + printed output |

**Comments**

- Good generic helper (`parse_insert`) used twice. Showing the table contents is a mark the reference solution omits; yours earns it.
- `INSERT INTO table VALUES (?,…)` relies on file column order matching table column order; naming the columns is safer. Calling `tables()` at the top of this cell silently wipes the database: say so in a comment.

## Task 4.3
A new table `Villa_Booking`, needs creating to store every date that each villa has been booked.
This will have two fields: one for the villa ID and one for the date.

For example, if villa number 1 is booked from the 01-Jan for 4 days, the following data will be inserted
into `Villa_Booking`:

```
1 01-Jan
1 02-Jan
1 03-Jan
1 04-Jan
```

Write a Python program that uses SQL to create `Villa_Booking` and populate it with the existing customer bookings. <div style="text-align: right">[5]</div>

In [19]:
# Task 4.3

from datetime import date,timedelta

villa_booking_data = []
for entry in customerbooking_data:
    villaID = entry[2]
    start_date = date.strptime(f"{entry[-2]}-2025","%d-%b-%Y")
    for day in range(int(entry[-1])):
        villa_booking_data.append((villaID,date.strftime(start_date,"%d-%b")))
        start_date += timedelta(days=1)

import sqlite3

with sqlite3.connect("holiday_company.db") as con:
    c = con.cursor()
    c.execute("""
    CREATE TABLE IF NOT EXISTS VillaBooking(
    villaID INTEGER REFERENCES Villa(villaID),
    occupationDate TEXT,
    PRIMARY KEY(villaID,occupationDate)
    )
    """)
    c.execute("DELETE FROM VillaBooking")
    c.executemany("INSERT INTO VillaBooking VALUES (?,?)",villa_booking_data)
    con.commit()



    


### 📝 MARKING — Task 4.3 Villa_Booking: **5 / 5**

| # | Marking point | Mark |
|---|---|---|
| 1 | Creates the Villa_Booking table | ✅ (`VillaBooking`, composite primary key) |
| 2 | Retrieves the existing booking data | ✅ from the list read in 4.2, not by a SELECT: accepted |
| 3 | Generates all booked dates for each booking | ✅ (verified on stand-in data: 08-Apr, 09-Apr, then 12-14 Apr) |
| 4 | Inserts each villa/date pair | ✅ |
| 5 | Commits the changes | ✅ |

**Comments**

- **Python version:** `date.strptime(...)` does **not exist on Python 3.7 – 3.13** (`date` has no `strptime`; only `datetime` does; `date.strptime` arrived in 3.14, I believe). Your saved outputs show the cell *did* run on your machine, so your interpreter there is much newer than 3.7. On 3.7 this line raises `AttributeError` and, because it comes before the `CREATE TABLE`, **nothing in the cell runs**. Portable fix: `from datetime import datetime, timedelta` and `datetime.strptime(...)`.
- Marked as run in your environment. If the exam machine is 3.7, treat up to 5 marks here as at risk.
- A table called `VillaBooking` where the paper says `Villa_Booking`: fine for the marks, but match the paper's name if a later task refers to it.
- Reading the bookings back with `SELECT villaID, startDate, numberOfDays FROM CustomerBooking` would make the cell independent of the earlier cell (it currently depends on `customerbooking_data`).

## Task 4.4
Write a Python function to:
- take as input the villa name, start date and number of days a customer would like to book
- use SQL to check the availability of the villa for the customer's request
- output a list of all the requested dates that the villa is available
- output a list of all the requested dates that the villa is not available.

All outputs must have an appropriate message. <div style="text-align: right">[6]</div>

Test your program with the following data:

Villa name      `Dolphin`

Month           `Apr`

Date            `8`

Number of days  `4`
<div style="text-align: right">[1]</div>

In [ ]:
def check_avail(villaName,start_date,number_of_days):
    import sqlite3

    with sqlite3.connect("holiday_company.db") as con:
        c = con.cursor()
        c.execute("SELECT villaID FROM Villa WHERE villaName = ?",(villaName,))
        villaID = c.fetchone()[0]
        
    requested_dates = []
    
    start_date = date.strptime(f"{start_date}-2025","%d-%b-%Y")
    for day in range(number_of_days):
        requested_dates.append((villaID,date.strftime(start_date,"%d-%b")))
        start_date += timedelta(days=1)

    avail_dict = {
        "available": [],
        "unavailable":[]
    }

    with sqlite3.connect("holiday_company.db") as con:
        c = con.cursor()
        for i in requested_dates:
            c.execute("SELECT * FROM VillaBooking WHERE VillaID = ? AND OccupationDate = ?",i)
            data = c.fetchone()
            if data is None:
                avail_dict["available"].append(i[1])
            else:
                avail_dict["unavailable"].append(i[1])
    print(f"""
Villa Name: {villaName}
Start Date: {start_date}
Number of Days: {number_of_days}

Available Dates:
{'\n'.join(["-"+i for i in avail_dict["available"]])}

Unavailable Dates:
{'\n'.join(["-"+i for i in avail_dict["unavailable"]])}

    """)
        
        

check_avail("Dolphin","08-Apr",4)
    


Villa Name: Dolphin
Start Date: 2025-04-12
Number of Days: 4

Available Dates:
-10-Apr
-11-Apr

Unavailable Dates:
-08-Apr
-09-Apr

    


### 📝 MARKING — Task 4.4 check_avail(): **5 / 7**

| # | Marking point | Mark |
|---|---|---|
| 1 | Accepts villa name, start date, number of days | ✅ |
| 2 | Generates all requested dates | ✅ (same `date.strptime` caveat as 4.3) |
| 3 | Uses SQL to find existing booked dates for the villa | ✅ |
| 4 | Correctly identifies dates that are **available** | ❌ inverted |
| 5 | Correctly identifies dates that are **unavailable** | ❌ inverted |
| 6 | Outputs both lists with appropriate messages | ✅ |
| 7 | Tests with Dolphin, 8 Apr, 4 days | ✅ |

**Comments**

- **Logic inverted.** `if data is None:` (no booking found = the date is **free**) appends to `"unavailable"`. On my stand-in data, with Dolphin booked on 08 and 09 April, your function reported 08-Apr and 09-Apr as *Available*. Your saved output ('Available: 10-Apr, 11-Apr; Unavailable: 08-Apr, 09-Apr') has the same swap. Reverse the two branches.
- **Wrong start date printed:** `Start Date: 2025-04-12` (saved output). You reuse the name `start_date` for the input string *and* for the running `datetime` that is advanced inside the loop, so by the time you print it, it has moved on. Keep the loop variable separate (`day`) or print the original parameter.
- **Python 3.7 SyntaxError:** the final `print(f"""…{'\n'.join(…)}…""")` has a **backslash inside an f-string expression**, which is illegal before Python 3.12. On 3.7 the whole cell fails to compile, so the function is never defined. Fix: `avail_text = "\n".join(...)` on its own line, then use `{avail_text}` in the f-string.
- `fetchone()[0]` crashes with `TypeError` for a villa name that isn't in the table; check for `None`.
- One SQL query per requested date is wasteful: one query for all booked dates of the villa, then `date in booked` per requested date (what the reference does), is shorter and easier to follow.
- `date` and `timedelta` come from the import in an earlier cell; import them inside this function or at the top of the notebook so the cell works on its own.